### ISSUE TITLE
AWS retry requests fail with 403 SignatureDoesNotMatch when newrelic distributed tracing is enabled.

### ISSUE THREAD

**Issue by andriiberezhynskyi** [newrelic/go-agent#427]
## Description

Every aws retry request fails with 403 SignatureDoesNotMatch when newrelic
distributed tracing is enabled.

## Steps to Reproduce

Here is a piece of code that reproduces the problem. In this example I inject
the request handler which forces first request attempt to always fail with
retry-able error (of course I don't do it in production, it's just to simulate
aws sdk request retries). Then the retried request fails with aws
SignatureDoesNotMatch error which indicates that newrelic agent has modified
retired request in such a why that it makes request's signature invalid.
How do I know it that the problem is caused by newrelic agent? If I remove the
line where I add newrelic txn to ctx then request is retried successfully.

```go
package main

import (
	"context"
	"log"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/request"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/s3/s3manager"
	"github.com/newrelic/go-agent/v3/integrations/nrawssdk-v1"
	"github.com/newrelic/go-agent/v3/newrelic"
)

func main() {
	s3bucket := "REDACTED"
	s3key := "REDACTED"
	nrLicenseKey := "REDACTED"
	nrAppName := "REDACTED"

	nrApp, err := newrelic.NewApplication(
		newrelic.ConfigAppName(nrAppName),
		newrelic.ConfigLicense(nrLicenseKey),

		func(config *newrelic.Config) {
			config.SpanEvents.Enabled = true
			config.DistributedTracer.Enabled = true
		},
	)
	if err != nil {
		log.Fatal(err)
	}

	if err := nrApp.WaitForConnection(5 * time.Second); nil != err {
		log.Fatal(err)
	}

	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))
	sess.Config.WithLogLevel(aws.LogDebugWithSigning | aws.LogDebugWithRequestErrors)

	nrawssdk.InstrumentHandlers(&sess.Handlers)

	// This forces the very first aws request to fail with retriable error.
	sess.Handlers.Send.SetBackNamed(request.NamedHandler{
		Name: "ForceFirstAttemptToFailWithRetriableError",
		Fn: func(r *request.Request) {
			if r.RetryCount == 0 {
				r.Error = &retryableError{}
			}
		},
	})

	txn := nrApp.StartTransaction("test_txn")
	defer txn.End()

	ctx := context.Background()

	// if you comment out the following line everything works fine: the
	// request is successfully retried by aws sdk
	ctx = newrelic.NewContext(ctx, txn)

	_, err = s3manager.NewDownloaderWithClient(s3.New(sess)).
		DownloadWithContext(ctx, aws.NewWriteAtBuffer([]byte{}), &s3.GetObjectInput{
			Bucket: aws.String(s3bucket),
			Key:    aws.String(s3key),
		})
	if err != nil {
		log.Fatal(err)
	}
}

type retryableError struct{}

func (e *retryableError) Error() string   { return "retry-able error" }
func (e *retryableError) Temporary() bool { return true }
```

## Expected Behavior

Newrelic agent does not make aws signature invalid when the aws request is
retried.

## Your Environment

Software versions from go.mod:

```
github.com/aws/aws-sdk-go v1.37.18
github.com/newrelic/go-agent/v3 v3.13.0
github.com/newrelic/go-agent/v3/integrations/nrawssdk-v1 v1.0.1
```

## Additional context

When I run the code example I posted above I see that the first request doesn't
have `newrelic,traceparent,tracestate` http headers. However retries have this
headers. My wild guess is that this additional headers that appear only in
retries make the aws signature invalid. You can see the logs [here](https://gist.github.com/andriiberezhynskyi/b3f472dddcfa348a768633d9ed534cf7).

This might seem like an edge case as this happens only if aws request is retried
but it turned out that aws s3 api returns retriable 500 http status code often
enough to make this problem quite annoying.

**Comment by rittneje** [comment]
This is because the New Relic headers are added in a "Send" handler, which the AWS SDK runs after the "Sign" handlers. So the order of events is as follows:
1. First, the AWS SDK signs the original request, without any New Relic headers. This algorithm incorporates header values into the signature. Since the New Relic headers are not set yet, they don't factor in.
2. Then, the New Relic headers get added.
3. Then, the request is sent to AWS.
4. If the first request fails and will be retried, the AWS SDK re-signs the request. This will incorporate the New Relic headers from the _first_ attempt.
5. Then, the New Relic headers get added again, with presumably new values.
6. Then, the second request is sent to AWS.
7. Since the signature was calculated with the old headers, AWS is not able to verify it with the new headers, so it fails.

There are two ways this can be fixed:
1. Have the New Relic headers get added in a "Sign" handler, _before_ the actual signature is generated. This will ensure the signature always includes the correct header values.
2. Have the New Relic headers removed in a "Retry" or "AfterRetry" handler. This will ensure the signature never includes the New Relic headers.

**Comment by rittneje** [comment]
The other thing I will mention is that realistically, there is no need for `InstrumentHandlers` to actually do anything for services than DynamoDB. This is because you can just configure the AWS SDK's HTTP client to use an instrumented RoundTripper via [`NewRoundTripper`](https://pkg.go.dev/github.com/newrelic/go-agent/v3@v3.15.2/newrelic#NewRoundTripper). Since this copies the request before adding the headers, the AWS SDK will never see them.

In other words, this `else` can probably just be removed.
https://github.com/newrelic/go-agent/blob/4b46fc5b40e5398695413097b5c461679a74eafd/_integrations/nrawssdk/internal/internal.go#L82-L84

**Comment by molecule-man** [comment]
> There are a few ways this can be fixed:
>
> 1.  Have the existing "EndNewRelicSegment" handler remove these headers. This will ensure the signature never includes the New Relic headers.
> 2. Have the New Relic headers get added in a "Sign" handler, before the actual signature is generated. This will ensure the signature always includes the correct header values.
> 3. Have the New Relic headers removed in a "Retry" or "AfterRetry" handler. This will ensure the signature never includes the New Relic headers.


@rittneje which solution is preferrable? I tried the solution 2 as it seemed the most straightforward and can confirm that it indeed fixes the issue: https://github.com/newrelic/go-agent/pull/429

**Comment by rittneje** [comment]
For me, removing that `else` block entirely (or perhaps just replacing `StartExternalSegment` with `StartSegment` so the span attributes have a place to go) would be the most preferable, since right now I think we may be double-counting our AWS requests. But if that is a backwards compatibility concern, then we may need to make a new function or a new major version or something.

I don't have a strong preference between the three original options I presented, but one thing to account for with (2) is that if the request is going to be pre-signed, then you cannot add any headers, because the user of the pre-signed URL would not expect to have to set those headers. That is, if `req.IsPresigned()`, then the handler should do nothing. It would definitely be good to add a unit test for this case: call `req.PresignRequest` and verify the New Relic headers are not in the resulting map.

**Comment by molecule-man** [comment]
@rittneje that makes sense. You are right the "Sign handler fix" breaks url presigning unless `req.IsPresigned()` is checked. Moreover it also causes SignatureDoesNotMatch if both `nrawssdk.InstrumentHandlers` and `newrelic.NewRoundTripper` are used.

**Comment by rittneje** [comment]
The latter could be fixed if `NewRoundTripper` did not inject the headers if they are already there, but that may be undesirable for other reasons.

I am still hoping one of the New Relic maintainers can weigh in on whether we can just remove that call to `StartExternalSegment`, as that would solve multiple problems at once.

**Comment by rittneje** [comment]
I also filed https://github.com/aws/aws-sdk-go/issues/4226 as this would simplify the problem for anyone injecting headers.

**Comment by rittneje** [comment]
Another question to consider is why New Relic is even inserting these headers, since AWS isn't going to use/forward them.

**Comment by RichVanderwal** [comment]
Hi @andriiberezhynskyi , we've just merged @rittneje 's [PR](https://github.com/newrelic/go-agent/pull/433) to our develop branch -- it sidesteps the issue altogether by not including headers for Distributed Tracing at all. We'd appreciate it if you tried it out and let us know your results!

**Comment by andriiberezhynskyi** [comment]
Hi @RichVanderwal 
I tried the develop branch out and confirm that @rittneje 's PR fixes the problem.

**Comment by Ak-x** [comment]
Closing as @rittneje 's PR resolves this issue.
